modules/collect/src/main/java/com/opengamma/strata/collect/io/XmlFile.java

```diff
@@ -0,0 +1,282 @@
+   * <p>
+   * This parses the specified byte source expecting an XML file format.
+   * The resulting instance can be queried for the root element.
+   * <p>
+   * This supports capturing attribute references, such as an id/href pair.
+   * Wherever the parser finds an attribute with the specified name, the element is added
+   * to the internal map, accessible by calling {@link #getReferences()}.
+   * 
+   * @param source  the XML source data
+   * @param refAttrName  the attribute name that should be parsed as a reference
+   * @return the parsed file
+   * @throws UncheckedIOException if an IO exception occurs
+   * @throws IllegalArgumentException if the file cannot be parsed
+   */
+  public static XmlFile of(ByteSource source, String refAttrName) {
+    ArgChecker.notNull(source, "source");
+    return Unchecked.wrap(() -> {
+      try (InputStream in = source.openBufferedStream()) {
+        XMLStreamReader xmlReader = xmlInputFactory().createXMLStreamReader(in);
+        try {
+          HashMap<String, XmlElement> refs = new HashMap<>();
+          XmlElement root = parse(xmlReader, refAttrName, refs);
+          return new XmlFile(root, refs);
+        } finally {
+          xmlReader.close();
+        }
+      }
+    });
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Parses the tree from the StAX stream reader, capturing references.
+   * <p>
+   * The reader should be created using the factory returned from {@link #xmlInputFactory()}.
+   * <p>
+   * This method supports capturing attribute references, such as an id/href pair.
+   * Wherever the parser finds an attribute with the specified name, the element is added
+   * to the specified map. Note that the map is mutated.
+   * 
+   * @param reader  the StAX stream reader, positioned at or before the element to be parsed
+   * @param refAttr  the attribute name that should be parsed as a reference, null if not applicable
+   * @param refs  the mutable map of references to update, null if not applicable
+   * @return the parsed element
+   * @throws IllegalArgumentException if the input cannot be parsed
+   */
+  private static XmlElement parse(XMLStreamReader reader, String refAttr, Map<String, XmlElement> refs) {
+    try {
+      // parse start element
+      String elementName = parseElementName(reader);
+      ImmutableMap<String, String> attrs = parseAttributes(reader);
+
+      // parse children or content
+      ImmutableList.Builder<XmlElement> childBuilder = ImmutableList.builder();
+      String content = "";
+      int event = reader.next();
+      while (event != XMLStreamConstants.END_ELEMENT) {
+        switch (event) {
+          case XMLStreamConstants.START_ELEMENT:
+            childBuilder.add(parse(reader, refAttr, refs));
+            break;
+          case XMLStreamConstants.CHARACTERS:
+          case XMLStreamConstants.CDATA:
+            content += reader.getText();
```

Comment:
Since `XMLStreamReader` has `IS_COALESCING=true` there should only be one call here.